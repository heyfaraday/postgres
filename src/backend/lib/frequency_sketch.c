#include "postgres.h"

#include <math.h>

#include "access/hash.h"
#include "lib/frequency_sketch.h"

#define MAX_HASH_FUNCS          10
#define BLOCK_SIZE              sizeof(uint64) * BITS_PER_BYTE
#define POWER_OF_BLOCK_SIZE     6

struct frequency_sketch
{
    int         k_hash_funcs;
    uint64      seed;

    int         bits_per_elem;
    uint64      elem_mask;
    uint64      reset_mask;
    uint64      size;

    uint64      bitset[FLEXIBLE_ARRAY_MEMBER];
};

static uint32 get_index(uint32 hash);
static uint32 get_offset(uint32 hash);
static int	my_sketch_power(uint64 target_bitset_bits);
static int	optimal_k(uint64 bitset_bits, int64 total_elems);
static void k_hashes(frequency_sketch *sketch, uint32 *hashes, unsigned char *elem,
                     size_t len);
static inline uint32 mod_m(uint32 a, uint64 m);

frequency_sketch *
sketch_create(int64 total_elems, int sketch_work_mem, uint64 seed, int bits_per_elem)
{
    frequency_sketch    *sketch;
    int         sketch_power;
    uint64      bitset_bytes;
    uint64      bitset_bits;
    uint64      bitset_elems;
    int         i;

    Assert((bits_per_elem - 1) & bits_per_elem == 0);

    bitset_bytes = Min(sketch_work_mem * UINT64CONST(1024), total_elems * bits_per_elem * 2);
    bitset_bytes = Max(1024 * 1024, bitset_bytes);

    sketch_power = my_sketch_power(bitset_bytes * BITS_PER_BYTE);
    bitset_bits = UINT64CONST(1) << sketch_power;
    bitset_bytes = bitset_bits / BITS_PER_BYTE;
    bitset_elems = bitset_bits / bits_per_elem;

    sketch = palloc0(offsetof(frequency_sketch, bitset) +
            sizeof(unsigned char) * bitset_bytes);
    sketch->k_hash_funcs = optimal_k(bitset_elems, total_elems);
    sketch->seed = seed;
    sketch->bits_per_elem = bits_per_elem;
    sketch->elem_mask = (UINT64CONST(1) << bits_per_elem) - 1;
    for (i = 0; i < BLOCK_SIZE / bits_per_elem; ++i)
    {
        sketch->reset_mask |= ((UINT64CONST(1) << (bits_per_elem - 1)) - 1) << bits_per_elem * i;
    }
    sketch->size = bitset_elems;

    return sketch;
}

void
sketch_free(frequency_sketch *sketch)
{
    pfree(sketch);
}

void
sketch_add_element(frequency_sketch *sketch, unsigned char *elem, size_t len)
{
    uint32  hashes[MAX_HASH_FUNCS];
    int     i;
    uint32  index;
    uint32  offset;
    uint64  min = UINT64CONST(1) << sketch->bits_per_elem;

    k_hashes(sketch, hashes, elem, len);

    for (i = 0; i < sketch->k_hash_funcs; ++i)
    {
        index = get_index(hashes[i]);
        offset = get_offset(hashes[i]);
        min = Min(min, sketch->bitset[index] >> offset & sketch->elem_mask);
    }

    if (min == sketch->elem_mask) return;

    for (i = 0; i < sketch->k_hash_funcs; ++i)
    {
        index = get_index(hashes[i]);
        offset = get_offset(hashes[i]);
        if ((sketch->bitset[index] >> offset & sketch->elem_mask) == min)
        {
            sketch->bitset[index] += UINT64CONST(1) << offset;
        }
    }
}

uint64
sketch_get_frequency(frequency_sketch *sketch, unsigned char *elem, size_t len)
{
    uint32  hashes[MAX_HASH_FUNCS];
    int     i;
    uint32  index;
    uint32  offset;
    uint64  min = UINT64CONST(1) << sketch->bits_per_elem;

    k_hashes(sketch, hashes, elem, len);

    for (i = 0; i < sketch->k_hash_funcs; ++i)
    {
        index = get_index(hashes[i]);
        offset = get_offset(hashes[i]);
        min = Min(min, sketch->bitset[index] >> offset & sketch->elem_mask);
    }

    return min;
}

void
sketch_reset(frequency_sketch *sketch)
{
    int i;

    for (i = 0; i < sketch->size; ++i) {
        sketch->bitset[i] >>= 1;
        sketch->bitset[i] &= sketch->reset_mask;
    }
}

static uint32
get_index(uint32 hash)
{
    return hash >> POWER_OF_BLOCK_SIZE;
}

static uint32
get_offset(uint32 hash)
{
    return hash & ((1 << POWER_OF_BLOCK_SIZE) - 1);
}

static int
my_sketch_power(uint64 target_bitset_bits)
{
    int			sketch_power = -1;

    while (target_bitset_bits > 0 && sketch_power < 32)
    {
        sketch_power++;
        target_bitset_bits >>= 1;
    }

    return sketch_power;
}

static int
optimal_k(uint64 bitset_bits, int64 total_elems)
{
    int			k = rint(log(2.0) * bitset_bits / total_elems);

    return Max(1, Min(k, MAX_HASH_FUNCS));
}

static void
k_hashes(frequency_sketch *filter, uint32 *hashes, unsigned char *elem, size_t len)
{
    uint64		hash;
    uint32		x;
    uint32      y;
    uint64		m;
    int			i;

    /* Use 64-bit hashing to get two independent 32-bit hashes */
    hash = DatumGetUInt64(hash_any_extended(elem, len, filter->seed));
    x = (uint32) hash;
    y = (uint32) (hash >> 32);
    m = filter->size;

    x = mod_m(x, m);
    y = mod_m(y, m);

    /* Accumulate hashes */
    hashes[0] = x * filter->bits_per_elem;
    for (i = 1; i < filter->k_hash_funcs; i++)
    {
        x = mod_m(x + y, m);
        y = mod_m(y + i, m);

        hashes[i] = x * filter->bits_per_elem;
    }
}

static inline uint32
mod_m(uint32 val, uint64 m)
{
    Assert(m <= PG_UINT32_MAX + UINT64CONST(1));
    Assert(((m - 1) & m) == 0);

    return val & (m - 1);
}

