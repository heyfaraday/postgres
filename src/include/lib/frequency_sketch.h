#ifndef FREQUENCY_SKETCH_H
#define FREQUENCY_SKETCH_H

typedef struct frequency_sketch frequency_sketch;

extern frequency_sketch *sketch_create(int64 total_elems, int sketch_work_mem,
        uint64 seed, int bits_per_elem);
extern void sketch_free(frequency_sketch *sketch);
extern uint64 sketch_used_mem(frequency_sketch *sketch);
		extern void sketch_add_element(frequency_sketch *sketch, unsigned char *elem,
        size_t len);
extern uint64 sketch_get_frequency(frequency_sketch *sketch,
        unsigned char *elem, size_t len);
extern void sketch_reset(frequency_sketch *sketch);


#endif                      /* FREQUENCY_SKETCH_H */
