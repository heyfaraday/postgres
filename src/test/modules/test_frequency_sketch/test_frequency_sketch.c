#include "postgres.h"

#include "access/hash.h"
#include "fmgr.h"
#include "lib/frequency_sketch.h"
#include "miscadmin.h"

PG_MODULE_MAGIC;

typedef struct for_data
{
	uint32 key;
	char status;
	uint64 value;
} for_data;

#define SH_PREFIX freq
#define SH_ELEMENT_TYPE for_data
#define SH_KEY_TYPE uint32
#define SH_KEY key
#define SH_HASH_KEY(tb, key) hash_uint32(key)
#define SH_EQUAL(tb, a, b) a == b
#define SH_SCOPE static inline
#define SH_DEFINE
#define SH_DECLARE

#include "lib/simplehash.h"

typedef uint64 (*rndptr_t)(int helper, int mod);

uint64 sum_rand(int helper, int mod)
{
	int i;
	uint64 rnd = 0;

	for (i = 0; i < helper; ++i)
	{
		rnd += random() % mod;
	}

	return rnd % mod;
}

uint64 mul_rand(int helper, int mod)
{
	int i;
	uint64 rnd = 1;

	for (i = 0; i < helper; ++i)
	{
		rnd *= random() % mod;
		rnd %= mod;
	}

	return rnd;
}

static void
create_and_test_sketch(int64 nelements, int bits_per_elem, int work_mem, int callerseed, int mod,
		rndptr_t rnd, int helper, const char *deviation)
{
	uint64		seed;
	uint64		tmp;
	uint64 		tmp2;
	uint64 		val1;
	uint64 		val2;
	uint64 		tries = 0;
	bool 		found;
	int64		errors = 0;
	frequency_sketch	*sketch;
	freq_hash	*hash_table;
	int 		i;
	for_data	*data;

	srandom(callerseed);

	seed = callerseed < 0 ? random() % PG_INT32_MAX : callerseed;

	hash_table = freq_create(CurrentMemoryContext, nelements, NULL);
	sketch = sketch_create(nelements, work_mem, seed, bits_per_elem);

	for (i = 0; i < 2 * nelements; ++i)
	{
		if (random() % 2)
		{
			tmp = rnd(helper, mod);
			sketch_add_element(sketch, (unsigned char *) &tmp, sizeof(tmp));
			++(freq_insert(hash_table, tmp, &found)->value);
		}
		else
		{
			tmp = rnd(helper, mod) % mod;
			tmp2 = rnd(helper, mod) % mod;
			data = freq_lookup(hash_table, tmp);
			val1 = (data ? data->value : 0);
			data = freq_lookup(hash_table, tmp2);
			val2 = (data ? data->value : 0);
			tmp = sketch_get_frequency(sketch, (unsigned char *) &tmp, sizeof(tmp));
			tmp2 = sketch_get_frequency(sketch, (unsigned char *) &tmp2, sizeof(tmp2));
			++tries;

			if ((val1 < val2 && tmp >= tmp2) || (val1 == val2 && tmp != tmp2) || (val1 > val2 && tmp <= tmp2))
			{
				++errors;
			}
		}
	}

	elog(DEBUG1, " FREQUENCY SKETCH:\nbits per element: %d\nused mem: %lu MB\nrandom distribution: %s\n"
			  "numbers bound: %d\nadded %ld elements\nhits: %ld\nmisses: %ld\n"
	          "hit ratio: %f\n  -----------------\n",
		   	  bits_per_elem, sketch_used_mem(sketch) >> 20, deviation, mod,
		   	  2 * nelements - tries, tries - errors, errors,
			  (double)(tries - errors) / tries);

	freq_destroy(hash_table);
	sketch_free(sketch);
}

PG_FUNCTION_INFO_V1(test_frequency_sketch);

/*
 * SQL-callable entry point to perform all tests.
 *
 * If a 1% false positive threshold is not met, emits WARNINGs.
 *
 * See README for details of arguments.
 */
Datum
test_frequency_sketch(PG_FUNCTION_ARGS)
{
	int64		nelements = PG_GETARG_INT64(0);
	int 		bits_per_elem = PG_GETARG_INT32(1);
	int 		work_mem = PG_GETARG_INT32(2);
	int			seed = PG_GETARG_INT32(3);
	int 		mod = PG_GETARG_INT32(4);
	int			tests = PG_GETARG_INT32(5);
	int			i;

	if (tests <= 0)
		elog(ERROR, "invalid number of tests: %d", tests);

	if (nelements < 0)
		elog(ERROR, "invalid number of elements: %d", nelements);

	if (((bits_per_elem - 1) & bits_per_elem) != 0 || bits_per_elem <= 0 || bits_per_elem > 64)
	{
		elog(ERROR, "invalid number of bits per element (should be a power of 2,"
			  "less or equal than 64: %d", bits_per_elem);
	}

	if (work_mem <= 0)
		elog(ERROR, "invalid number of work mem: %d", work_mem);

	if (mod <= 0)
		elog(ERROR, "invalid mod: %d", mod);

	for (i = 0; i < tests; i++)
	{
		elog(DEBUG1, "beginning test case #%d...", i + 1);

		create_and_test_sketch(nelements, bits_per_elem, work_mem, seed, mod, sum_rand, 1, "~uniform");
		create_and_test_sketch(nelements, bits_per_elem, work_mem, seed, mod, sum_rand, 10,
				"almost normal");
		create_and_test_sketch(nelements, bits_per_elem, work_mem, seed, mod, mul_rand, 2, "~uniform squared");
		create_and_test_sketch(nelements, bits_per_elem, work_mem, seed, mod, mul_rand, 3, "~uniform cubed");
	}

	PG_RETURN_VOID();
}
