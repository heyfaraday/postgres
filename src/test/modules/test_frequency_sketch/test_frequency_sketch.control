comment = 'Test code for Frequency sketch library'
default_version = '1.0'
module_pathname = '$libdir/test_frequency_sketch'
relocatable = true
