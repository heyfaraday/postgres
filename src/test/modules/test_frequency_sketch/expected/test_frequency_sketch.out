CREATE EXTENSION test_frequency_sketch;
-- See README for explanation of arguments:
SELECT test_frequency_sketch(nelements => 1000000,
                      bits_per_elem => 4,
                      work_mem => 1073741824,
                      seed => 11111,
                      mod => 100000,
                      tests => 1);
 test_frequency_sketch 
-----------------------
 
(1 row)

SELECT test_frequency_sketch(nelements => 100000,
                             bits_per_elem => 4,
                             work_mem => 1073741824,
                             seed => 11111,
                             mod => 100000,
                             tests => 1);
 test_frequency_sketch 
-----------------------
 
(1 row)

SELECT test_frequency_sketch(nelements => 1000000,
                             bits_per_elem => 8,
                             work_mem => 1073741824,
                             seed => 11111,
                             mod => 100000,
                             tests => 1);
 test_frequency_sketch 
-----------------------
 
(1 row)

SELECT test_frequency_sketch(nelements => 1000000,
                             bits_per_elem => 16,
                             work_mem => 1073741824,
                             seed => 11111,
                             mod => 100,
                             tests => 1);
 test_frequency_sketch 
-----------------------
 
(1 row)

